src/computability/NFA.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+-/
+
+import data.fintype.basic
+import computability.DFA
+
+/-!
+# Nondeterministic Finite Automata
+This file contains the definition of a Nondeterministic Finite Automaton (NFA), a state machine
+which determines whether a string (implemented as a list over an arbitrary alphabet) is in a regular
+set by evaluating the string over every possible path.
+We show that DFA's are equivalent to NFA's however the construction from NFA to DFA uses an
+exponential number of states.
+-/
+
+universes u v
+
+/-- An NFA is a set of states (`state`), a transition function from state to state labelled by the
+  alphabet (`step`), a starting state (`start`) and a set of acceptance states (`accept`).
+  Note the transition function sends a state to a `finset` of states. These are the states that it
+  may be sent to. -/
+structure NFA (alphabet : Type u) (state : Type v) :=
+(step : state → alphabet → finset state)
+(start : finset state)
+(accept : finset state)
+
+namespace NFA
+
+variables {α : Type u} {σ σ' σ₁ σ₂ σ₃ : Type v} (M : NFA α σ)
+variables [decidable_eq σ] [decidable_eq σ₁] [decidable_eq σ₂] [decidable_eq σ₃]
+
+instance NFA_inhabited : inhabited (NFA α σ') := ⟨ NFA.mk (λ _ _, ∅) ∅ ∅ ⟩
```

Comment:
```suggestion
instance : inhabited (NFA α σ') := ⟨ NFA.mk (λ _ _, ∅) ∅ ∅ ⟩
```
this gets the name `NFA_inhabited` automatically, since it's in the `NFA` namespace